src/CMakeLists.txt

```
@@ -0,0 +1,75 @@
+add_library(minisketch_field_sources INTERFACE)
+target_sources(minisketch_field_sources
+  INTERFACE
+    fields/generic_1byte.cpp
+    fields/generic_2bytes.cpp
+    fields/generic_3bytes.cpp
+    fields/generic_4bytes.cpp
+    fields/generic_5bytes.cpp
+    fields/generic_6bytes.cpp
+    fields/generic_7bytes.cpp
+    fields/generic_8bytes.cpp
+)
+if(HAVE_CLMUL)
+  set(clmul_sources
+    fields/clmul_1byte.cpp
+    fields/clmul_2bytes.cpp
+    fields/clmul_3bytes.cpp
+    fields/clmul_4bytes.cpp
+    fields/clmul_5bytes.cpp
+    fields/clmul_6bytes.cpp
+    fields/clmul_7bytes.cpp
+    fields/clmul_8bytes.cpp
+  )
+  set_property(SOURCE minisketch.cpp
+    PROPERTY OMPILE_DEFINITIONS HAVE_CLMUL
+  )
+  set_property(SOURCE ${clmul_sources}
+    PROPERTY COMPILE_OPTIONS ${CLMUL_CXXFLAGS}
+  )
+  target_sources(minisketch_field_sources INTERFACE ${clmul_sources})
+endif()
+
+add_compile_definitions($<$<BOOL:${HAVE_CLZ}>:HAVE_CLZ>)
+
+add_library(minisketch minisketch.cpp)
+set_property(TARGET minisketch PROPERTY PUBLIC_HEADER
+  ${PROJECT_SOURCE_DIR}/include/minisketch.h
+)
+target_include_directories(minisketch INTERFACE
+  $<BUILD_INTERFACE:$<$<NOT:$<BOOL:${PROJECT_IS_TOP_LEVEL}>>:${PROJECT_SOURCE_DIR}/include>>
+)
```

Is the include directory already set when this is the top level and you only want to set it otherwise?

```suggestion
target_include_directories(minisketch PUBLIC
  $<BUILD_INTERFACE:${minisketch_SOURCE_DIR}/include>
)
```

---

> Is the include directory already set when this is the top level...

This is not necessary, because project files use relative paths for includes, for example, `#include "../include/minisketch.h"`.

> and you only want to set it otherwise?

Yes. Only `INTERFACE_INCLUDE_DIRECTORIES` is necessary as a usage requirement.

---

That explains the rationale for `INTERFACE` (even though I would prefer `PUBLIC` and avoid relative includes), but it does not explain the rationale for the `$<NOT:$<BOOL:${PROJECT_IS_TOP_LEVEL}>>` condition.

---

> That explains the rationale for `INTERFACE` (even though I would prefer `PUBLIC` and avoid relative includes)...

This idea deserves its own discussion in a separated issues or PR.

---

>  it does not explain the rationale for the `$<NOT:$<BOOL:${PROJECT_IS_TOP_LEVEL}>>` condition.

Thanks! Dropped.

---

> > I would prefer `PUBLIC` and avoid relative includes...
> 
> This idea deserves its own discussion in a separated issues or PR.

Keeping this as `INTERFACE` prevents its inclusion (as-intended) when building the internal objects.

Edit: Ah, I see that you understood that. The relative includes guarantee that we're never accidentally picking up system headers.